File: grpc_memleak/randnum_client.py
from __future__ import print_function
import grpc

import randnum_pb2
import randnum_pb2_grpc

def run():
  channel = grpc.insecure_channel('localhost:50051')
  stub = randnum_pb2_grpc.RandomNumbersStub(channel)
  gen = stub.GetNext(randnum_pb2.EmptyRequest())
  count = 0
  step = 0
  for num in gen:
    if count % 10000 == 0:
      print(num.value)
      count = 0
    if step == 20000:
      break
    count += 1
    step += 1

if __name__ == '__main__':
  run()

File: grpc_memleak/randnum_client.cc
#include <chrono>
#include <iostream>
#include <memory>
#include <random>
#include <string>
#include <thread>

#include <grpc/grpc.h>
#include <grpc++/channel.h>
#include <grpc++/client_context.h>
#include <grpc++/create_channel.h>
#include <grpc++/security/credentials.h>
#include "randnum.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::ClientReader;
using grpc::Status;
using randnum::EmptyRequest;
using randnum::Number;
using randnum::RandomNumbers;

class RandomNumbersClient {
 public:
  RandomNumbersClient(std::shared_ptr<Channel> channel)
      : stub_(RandomNumbers::NewStub(channel)) {} 

  void GetNumbers() {
    Number num;
    ClientContext context;
    EmptyRequest req;

    std::unique_ptr<ClientReader<Number> > reader(
        stub_->GetNext(&context, req));
    int count = 0;
    while (reader->Read(&num)) {
      if(count % 10000 == 0)
      {
        std::cout << num.value() << std::endl;
        count = 0;
      } 
      ++count;
    }
  }

 private:
  std::unique_ptr<RandomNumbers::Stub> stub_;
};

int main(int argc, char** argv) {
  RandomNumbersClient client(
    grpc::CreateChannel("localhost:50051", 
    grpc::InsecureChannelCredentials()));
  client.GetNumbers();
  return 0;
}
